ocaml/runtime/extern.c

```diff
@@ -1279,8 +1240,26 @@ intnat reachable_words_once(value root, enum reachable_words_traversal_mode mode
         } else {
           extern_update_location_with_data(h, new_mark);
         }
+
         /* The block contributes to the total size */
         size += 1 + sz;           /* header word included */
+        if (sizes_by_root_id) {
+          if (new_mark == Shared) {
+#pragma GCC diagnostic push
+            /* GCC complains that [mark] could be uninitialized. This can only happen
+             * if previously_marked returns false. But then new_mark is set to
+             * identifier which by assumption is != Shared. */
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+            /* mark is identifier of some other root that we counted this node
+             * as contributing to. Since it is evidently not uniquely reachable, we
+             * undo this contribution */
+            /* Need to shift left by 1 to respect the OCaml representation of [int] values */
+            Field(sizes_by_root_id, mark) -= (1 + sz) << 1;
+#pragma GCC diagnostic pop
+          } else {
```

Comment:
The possibilities inside this `else` are `new_mark == root` and `new_mark == RootProcessed`, no? Worth asserting `new_mark == root || new_mark == RootProcessed`, I think, since it's hard to keep track when reading

Reply:
I think you meant `new_mark == identifier` (root is a value which has a different type). I've added an assert for this.